src/nupic/algorithms/sdr_classifier.py

```diff
@@ -0,0 +1,354 @@
+# it under the terms of the GNU Affero Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Affero Public License for more details.
+#
+# You should have received a copy of the GNU Affero Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+"""
+Implementation for a SDR classifier.
+
+The SDR classifier takes the form of a single layer classification network
+that takes SDRs as input and outputs a predicted distribution of classes.
+"""
+
+from collections import deque
+import itertools
+
+import numpy
+
+g_debugPrefix = "SDRClassifier"
+
+
+def _pFormatArray(array_, fmt="%.2f"):
+  """Return a string with pretty-print of a numpy array using the given format
+  for each element"""
+  return "[ " + " ".join(fmt % x for x in array_) + " ]"
+
+
+class SDRClassifier(object):
+  """
+  The SDR Classifier accepts a binary input pattern from the
+  level below (the "activationPattern") and information from the sensor and
+  encoders (the "classification") describing the true (target) input.
+
+  The SDR classifier maps input patterns to class labels. There are as many
+  output units as the number of class labels (buckets). The output is a
+  probabilistic distribution over all class labels
+
+  During inference, the output is calculated by first doing a weighted summation
+  of all the inputs, and then perform a softmax nonlinear function to get
+  the predicted distribution of class labels
+
+  During learning, the connection weights between input units and output units
+  are adjusted to maximize the likelihood of the model
+
+  References:
+    Alex Graves. Supervised Sequence Labeling with Recurrent Neural Networks
+    PhD Thesis, 2008
+
+    J. S. Bridle. Probabilistic interpretation of feedforward classification
+    network outputs, with relationships to statistical pattern recognition.
+    In F. Fogleman-Soulie and J.Herault, editors, Neurocomputing: Algorithms,
+    Architectures and Applications, pp 227-236, Springer-Verlag, 1990
+  """
+
+  __VERSION__ = 1
+
+
+  def __init__(self,
+               steps=(1,),
+               alpha=0.001,
+               actValueAlpha=0.3,
+               verbosity=0):
+    """Constructor for the SDR classifier.
+
+    Parameters:
+    ---------------------------------------------------------------------
+    @param numInputs (int) Length of the input activation pattern
+    @param steps (list) Sequence of the different steps of multi-step
+        predictions to learn
+    @param alpha (float) The alpha used to adapt the weight matrix during
+        learning. A lower alpha results in longer term memory.
```

Comment:
Does alpha really control how long the memory is stored? (If an SDR is not seen for a really long time it will not forget it regardless of alpha.) It seems like alpha just controls how fast it adapts?